File: benchmarks/primes_parallel/c/primes_parallel.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

#define NUM_THREADS 4

typedef struct {
    long start, end;
    long count;
} ThreadArg;

int is_prime(long x) {
    if (x < 2) return 0;
    if (x % 2 == 0) return x == 2;
    for (long d = 3; d * d <= x; d += 2) {
        if (x % d == 0) return 0;
    }
    return 1;
}

void *worker(void *arg) {
    ThreadArg *t = (ThreadArg *)arg;
    long count = 0;
    for (long x = t->start; x < t->end; x++) {
        if (is_prime(x)) count++;
    }
    t->count = count;
    return NULL;
}

int main(int argc, char **argv) {
    long n = argc > 1 ? atol(argv[1]) : 2000000;

    if (!is_prime(2) || !is_prime(97) || is_prime(100) || is_prime(1)) {
        fprintf(stderr, "self-check failed: is_prime disagrees with known facts\n");
        return 1;
    }

    pthread_t threads[NUM_THREADS];
    ThreadArg args[NUM_THREADS];
    long chunk = n / NUM_THREADS;
    for (int i = 0; i < NUM_THREADS; i++) {
        args[i].start = i * chunk;
        if (args[i].start < 2) args[i].start = 2;
        args[i].end = (i == NUM_THREADS - 1) ? n : (i + 1) * chunk;
        pthread_create(&threads[i], NULL, worker, &args[i]);
    }

    long total = 0;
    for (int i = 0; i < NUM_THREADS; i++) {
        pthread_join(threads[i], NULL);
        total += args[i].count;
    }

    printf("%ld\n", total);
    return 0;
}

File: benchmarks/primes_parallel/python/primes_parallel.py
import sys
from multiprocessing import Pool

NUM_THREADS = 4


def is_prime(x):
    if x < 2:
        return False
    if x % 2 == 0:
        return x == 2
    d = 3
    while d * d <= x:
        if x % d == 0:
            return False
        d += 2
    return True


def count_range(bounds):
    start, end = bounds
    return sum(1 for x in range(start, end) if is_prime(x))


def main():
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 2_000_000

    assert is_prime(2) and is_prime(97) and not is_prime(100) and not is_prime(1), \
        "self-check failed: is_prime disagrees with known facts"

    chunk = n // NUM_THREADS
    ranges = []
    for i in range(NUM_THREADS):
        start = max(i * chunk, 2)
        end = n if i == NUM_THREADS - 1 else (i + 1) * chunk
        ranges.append((start, end))

    with Pool(NUM_THREADS) as pool:
        total = sum(pool.map(count_range, ranges))

    print(total)


if __name__ == "__main__":
    main()
